### ISSUE TITLE
Cannot type solve lambdas in Stream's forEach

### ISSUE THREAD

**Issue by JBrVJxsc** [javaparser/javaparser#2160]
Version: 5.3.1

Hey guys, here's the testing code from JavaParser's example repo: https://github.com/javaparser/javaparser-visited/blob/master/src/main/java/org/javaparser/samples/ReversePolishNotation.java

And I'm trying to use the following code from the book to test symbol resolver:
```
public class GetTypeOfReference {
    private static final String FILE_PATH = "/src/main/java/org/javaparser/samples/ReversePolishNotation.java";

    public static void main(String[] args) throws Exception {
        TypeSolver typeSolver = new CombinedTypeSolver();
        JavaSymbolSolver symbolSolver = new JavaSymbolSolver(typeSolver);
        StaticJavaParser.getConfiguration().setSymbolResolver(symbolSolver);

        CompilationUnit cu = StaticJavaParser.parse(new File(FILE_PATH));
        cu.findAll(AssignExpr.class).forEach(ae -> {
            ResolvedType resolvedType = ae.calculateResolvedType();
            System.out.println(ae.toString() + " is a: " + resolvedType);
        });
    }
}
```

And i'm getting the following exceptions:
```
Exception in thread "main" java.lang.RuntimeException: Error calculating the type of parameter t -> {
    double a;
    double b;
    switch(t) {
        case "+":
            b = numbers.pop();
            a = numbers.pop();
            numbers.push(a + b);
            break;
        case "/":
            b = numbers.pop();
            a = numbers.pop();
            numbers.push(a / b);
            break;
        case "-":
            b = numbers.pop();
            a = numbers.pop();
            numbers.push(a - b);
            break;
        case "*":
            b = numbers.pop();
            a = numbers.pop();
            numbers.push(a * b);
            break;
        default:
            numbers.push(Double.valueOf(t));
    }
} of method call Stream.of(tokens).forEach(t -> {
    double a;
    double b;
    switch(t) {
        case "+":
            b = numbers.pop();
            a = numbers.pop();
            numbers.push(a + b);
            break;
        case "/":
            b = numbers.pop();
            a = numbers.pop();
            numbers.push(a / b);
            break;
        case "-":
            b = numbers.pop();
            a = numbers.pop();
            numbers.push(a - b);
            break;
        case "*":
            b = numbers.pop();
            a = numbers.pop();
            numbers.push(a * b);
            break;
        default:
            numbers.push(Double.valueOf(t));
    }
})
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.solveMethodAsUsage(JavaParserFacade.java:578)
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.LambdaExprContext.solveSymbolAsValue(LambdaExprContext.java:65)
	at com.github.javaparser.symbolsolver.javaparsermodel.declarations.JavaParserParameterDeclaration.getType(JavaParserParameterDeclaration.java:73)
	at com.github.javaparser.symbolsolver.model.resolution.Value.from(Value.java:40)
	at com.github.javaparser.symbolsolver.core.resolution.Context.solveSymbolAsValue(Context.java:67)
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.StatementContext.solveSymbolAsValue(StatementContext.java:143)
	at com.github.javaparser.symbolsolver.resolution.SymbolSolver.solveSymbolAsValue(SymbolSolver.java:68)
	at com.github.javaparser.symbolsolver.resolution.SymbolSolver.solveSymbolAsValue(SymbolSolver.java:73)
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:277)
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:44)
	at com.github.javaparser.ast.expr.NameExpr.accept(NameExpr.java:79)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getTypeConcrete(JavaParserFacade.java:448)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getType(JavaParserFacade.java:310)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getType(JavaParserFacade.java:292)
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.SwitchEntryContext.solveSymbol(SwitchEntryContext.java:48)
	at com.github.javaparser.symbolsolver.core.resolution.Context.solveSymbolAsValue(Context.java:65)
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.StatementContext.solveSymbolAsValue(StatementContext.java:143)
	at com.github.javaparser.symbolsolver.resolution.SymbolSolver.solveSymbolAsValue(SymbolSolver.java:68)
	at com.github.javaparser.symbolsolver.resolution.SymbolSolver.solveSymbolAsValue(SymbolSolver.java:73)
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:277)
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:44)
	at com.github.javaparser.ast.expr.NameExpr.accept(NameExpr.java:79)
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:98)
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:44)
	at com.github.javaparser.ast.expr.AssignExpr.accept(AssignExpr.java:135)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getTypeConcrete(JavaParserFacade.java:448)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getType(JavaParserFacade.java:310)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getType(JavaParserFacade.java:292)
	at com.github.javaparser.symbolsolver.JavaSymbolSolver.calculateType(JavaSymbolSolver.java:250)
	at com.github.javaparser.ast.expr.Expression.calculateResolvedType(Expression.java:564)
	at GetTypeOfReference.lambda$main$0(GetTypeOfReference.java:21)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
	at GetTypeOfReference.main(GetTypeOfReference.java:20)
Caused by: UnsolvedSymbolException{context='Stream.of(tokens).forEach(t -> {
    double a;
    double b;
    switch(t) {
        case "+":
            b = numbers.pop();
            a = numbers.pop();
            numbers.push(a + b);
            break;
        case "/":
            b = numbers.pop();
            a = numbers.pop();
            numbers.push(a / b);
            break;
        case "-":
            b = numbers.pop();
            a = numbers.pop();
            numbers.push(a - b);
            break;
        case "*":
            b = numbers.pop();
            a = numbers.pop();
            numbers.push(a * b);
            break;
        default:
            numbers.push(Double.valueOf(t));
    }
})', name='Stream.of(tokens)', cause='java.lang.RuntimeException: Error calculating the type of parameter tokens of method call Stream.of(tokens)'}
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.AbstractJavaParserContext.findTypeDeclarations(AbstractJavaParserContext.java:176)
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.MethodCallExprContext.solveMethod(MethodCallExprContext.java:152)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.solve(JavaParserFacade.java:242)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.solve(JavaParserFacade.java:125)
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:356)
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:44)
	at com.github.javaparser.ast.expr.LambdaExpr.accept(LambdaExpr.java:162)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getTypeConcrete(JavaParserFacade.java:448)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getType(JavaParserFacade.java:340)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.solveMethodAsUsage(JavaParserFacade.java:576)
	... 32 more
Caused by: java.lang.RuntimeException: Error calculating the type of parameter tokens of method call Stream.of(tokens)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.solveMethodAsUsage(JavaParserFacade.java:578)
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:267)
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:44)
	at com.github.javaparser.ast.expr.MethodCallExpr.accept(MethodCallExpr.java:115)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getTypeConcrete(JavaParserFacade.java:448)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getType(JavaParserFacade.java:310)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getType(JavaParserFacade.java:292)
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.AbstractJavaParserContext.findTypeDeclarations(AbstractJavaParserContext.java:167)
	... 41 more
Caused by: UnsolvedSymbolException{context='null', name='String', cause='null'}
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.convertToUsage(JavaParserFacade.java:509)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.convertToUsage(JavaParserFacade.java:543)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.convert(JavaParserFacade.java:567)
	at com.github.javaparser.symbolsolver.javaparsermodel.declarations.JavaParserSymbolDeclaration.getType(JavaParserSymbolDeclaration.java:146)
	at com.github.javaparser.symbolsolver.model.resolution.Value.from(Value.java:40)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.LinkedList$LLSpliterator.tryAdvance(LinkedList.java:1253)
	at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:127)
	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:502)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:488)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:543)
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.AbstractJavaParserContext.solveWithAsValue(AbstractJavaParserContext.java:149)
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.StatementContext.solveSymbolAsValue(StatementContext.java:135)
	at com.github.javaparser.symbolsolver.javaparsermodel.contexts.MethodCallExprContext.solveSymbolAsValue(MethodCallExprContext.java:147)
	at com.github.javaparser.symbolsolver.resolution.SymbolSolver.solveSymbolAsValue(SymbolSolver.java:68)
	at com.github.javaparser.symbolsolver.resolution.SymbolSolver.solveSymbolAsValue(SymbolSolver.java:73)
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:277)
	at com.github.javaparser.symbolsolver.javaparsermodel.TypeExtractor.visit(TypeExtractor.java:44)
	at com.github.javaparser.ast.expr.NameExpr.accept(NameExpr.java:79)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getTypeConcrete(JavaParserFacade.java:448)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.getType(JavaParserFacade.java:340)
	at com.github.javaparser.symbolsolver.javaparsermodel.JavaParserFacade.solveMethodAsUsage(JavaParserFacade.java:576)
	... 48 more
```

Thanks.

**Comment by matozoid** [comment]
Thanks! The best way to get a bug on the symbol solver fixed, is by finding the problem in the symbol solver code. Otherwise it might take a long time.

**Comment by villadora** [comment]
I just found some place would lead to this problem. 

https://github.com/javaparser/javaparser/blob/a0a883c87b50caf3a3913f8122a3dfe0037d0292/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/JavaParserFacade.java#L319

here should get the lambda expression type itself rather than get its' return type after evaluated.

Take for example:

```java
list.forEach(item-> {});
```

when getting type or arg `item->{}` should return `Consumer<T>` rather than `ResolvedVoidType`.

the `forEach` signature should be `Iterable.forEach(Consumer)` rather than `Iterable.forEach(Void)`.


ps:

https://github.com/javaparser/javaparser/blob/f2b2118ace92c8f0f8033449a54ab93c6a5984e5/javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/TypeExtractor.java#L430
there just return VoidType which cause the arg type in `forEach` become void.

**Comment by matozoid** [comment]
Let's start with asking @ftomassetti why the code currently works as it does. Why does the lambda resolve to `ResolvedVoidType`, not `Consumer<T>`? And will changing this break the whole world?

**Comment by matozoid** [comment]
@ftomassetti ?

**Comment by ftomassetti** [comment]
The comment from @villadora makes sense. Hopefully fixing it should not break the whole world. Lambdas & inferred types are imho the most difficult bit of the symbol solver to get right

**Comment by jlerbsc** [comment]
It seems to be corrected in version 3.14.4. Test output is

    b = numbers.pop() is a: PrimitiveTypeUsage{name='double'}
    a = numbers.pop() is a: PrimitiveTypeUsage{name='double'}
    b = numbers.pop() is a: PrimitiveTypeUsage{name='double'}
    a = numbers.pop() is a: PrimitiveTypeUsage{name='double'}
    b = numbers.pop() is a: PrimitiveTypeUsage{name='double'}
    a = numbers.pop() is a: PrimitiveTypeUsage{name='double'}
    b = numbers.pop() is a: PrimitiveTypeUsage{name='double'}
    a = numbers.pop() is a: PrimitiveTypeUsage{name='double'}
    memory = 0 is a: PrimitiveTypeUsage{name='double'}
    memory = value is a: PrimitiveTypeUsage{name='double'}

**Comment by matozoid** [comment]
@laonawuli are you available for a retest?
